### PR TITLE
Improve SPI Reliability

### PR DESCRIPTION
**Pull request by jaketenten**

### CODE REVIEW

File: user_modules/src/test.c [hunk 1]
```diff
@@ -149,10 +169,13 @@ static int spi_assemble_button_release(IEBUS_FRAME_INTERNAL *iebus_frame_interna
 
 int main(void)
 {
-#if 0
+#if 1
```
Comment (joseargao) [inline user_modules/src/test.c]: TODO: CircleCI?

File: user_modules/src/spi.c [hunk 2]
```diff
@@ -26,21 +26,27 @@
...
+    volatile bool transfer_thread_cancel_flag;   /* Receive cancel flag.
                                                  * Volatile to make the compiler mark it as illegal
                                                  * to optimize this out of store */
-    pthread_mutex_t lock;                       /* Lock for multithread sanity */
-    pthread_t receive_thread;                   /* SPI receive thread */
+    uint8_t spi_rx[SPI_BUFFER_SIZE];            /* RX Buffer */
+    uint8_t spi_tx[SPI_BUFFER_SIZE];            /* TX Buffer */
+    int spi_tx_idx;                             /* Current TX Buffer index, to tell where we are */
+    sem_t sem;                                  /* Semaphore for buffer availability */
+    sem_t sem2;                                 /* Semaphore for buffer availability */
```
[the remark this answers is not in the document]
Comment (jaketenten) [reply user_modules/src/spi.c]: Actually sem2 isn't used at all. Will remove sem2.

File: user_modules/src/spi.c [hunk 3]
```diff
@@ -90,27 +83,54 @@ static inline void teardown(void)
...
         int s;
-        pthread_join(spi_data.receive_thread, (void *)&s);
+        pthread_join(spi_data.transfer_thread, (void *)&s);
     }
 
     spi_data.on_data_received = NULL;
 }
 
-static inline bool transfer_one_byte(uint8_t *rx, const uint8_t *tx)
+static inline void hex_dump(const uint8_t *src, size_t length, size_t line_size,
```
[the remark this answers is not in the document]
Comment (jaketenten) [reply user_modules/src/spi.c]: True. This will generate a warning since it won't be used

File: user_modules/src/test.c (hunk 1 above)
Comment (jaketenten) [reply user_modules/src/test.c]: added a note

**Pull request merged**